File: ureka-framework-main/ureka_framework/model/message_model/message.py
# Data Model (Message)
from typing import Optional
from pydantic import BaseModel, ConfigDict, ValidationError


######################################################
# Message Operation
######################################################
MESSAGE_RECV_AND_STORE: str = "MESSAGE_RECV_AND_STORE"
MESSAGE_VERIFY_AND_EXECUTE: str = "MESSAGE_VERIFY_AND_EXECUTE"


######################################################
# Data Model
######################################################
class Message(BaseModel):
    # Message
    message_operation: Optional[str] = None
    message_type: Optional[str] = None
    message_str: Optional[str] = None

    # By default, Pydantic "ignore" extra input fields not defined in model schema
    # Moreover, we can explicitly "allow" or "forbid (with Error)" extra input fields not defined in model schema
    model_config = ConfigDict(extra="forbid")


################################################################################
#                                < Message_obj >                               #
#                                       | self-defined serilaization           #
#                                       | (all str, which is native type)      #
#                                       v                                      #
#         < JSON_dict (Should be JSON serializable, i.e. native type) >        #
#                                       |                                      #
#                                       v                                      #
#                       < JSON_str (Printable Characters) >                    #
################################################################################
def message_to_jsonstr(message_obj: Message) -> str:
    # "indent" do not affect json validation, but may affect json size!?
    message_json = message_obj.model_dump_json(indent=4, exclude_none=True)
    return message_json


def jsonstr_to_message(json_str: str) -> Message:
    try:
        return Message.model_validate_json(json_str)
    except ValidationError as error:
        failure_msg = "NOT VALID JSON or VALID MESSAGE SCHEMA"
        # simple_log("error", f"{failure_msg}: {error}")
        raise RuntimeError(failure_msg)

File: ureka-framework-java/app/src/main/java/ureka/framework/model/message_model/Message.java
package ureka.framework.model.message_model;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.util.Map;

public class Message {
    //////////////////////////////////////////////////////
    // Message Operation
    //////////////////////////////////////////////////////
    public static final String MESSAGE_RECV_AND_STORE = "MESSAGE_RECV_AND_STORE";
    public static final String MESSAGE_VERIFY_AND_EXECUTE = "MESSAGE_VERIFY_AND_EXECUTE";

    private String messageOperation;
    private String messageType;
    private String messageStr;

    public Message(Map<String, String> values) {
        this.messageOperation = values.getOrDefault("messageOperation", null);
        this.messageType = values.getOrDefault("messageType", null);
        this.messageStr = values.getOrDefault("messageStr", null);
    }

    public String getMessageOperation() {
        return messageOperation;
    }

    public void setMessageOperation(String messageOperation) {
        this.messageOperation = messageOperation;
    }

    public String getMessageType() {
        return messageType;
    }

    public void setMessageType(String messageType) {
        this.messageType = messageType;
    }

    public String getMessageStr() {
        return messageStr;
    }

    public void setMessageStr(String messageStr) {
        this.messageStr = messageStr;
    }

    public static String messageToJsonstr(Message messageObj) {
        Gson gson = new GsonBuilder().serializeNulls().create();
        return gson.toJson(messageObj);
    }

    public static Message jsonstrToMessage(String json) {
        Gson gson = new GsonBuilder().create();
        try {
            return gson.fromJson(json, Message.class);
        } catch (Exception e) {
            String failureMsg = "NOT VALID JSON or VALID MESSAGE SCHEMA";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw new RuntimeException(failureMsg);
        }
    }
}
